src/AVLTreeImp.h

```diff
@@ -0,0 +1,345 @@
+shared_ptr<TNode<Key, Info>> AVLTree<Key, Info>::getRoot() const {
+    return this->root;
+}
+
+template<class Key, class Info>
+shared_ptr<TNode<Key, Info>> AVLTree<Key, Info>::find(const Key &key) const {
+    return searchBinaryTree(key, this->root);
+}
+
+template<class Key, class Info>
+void AVLTree<Key, Info>::insert(const Key &key, const Info &info) {
+    shared_ptr<TNode<Key, Info>> newNode = make_shared<TNode<Key, Info>>();
+
+    newNode->setKey(key);
+    newNode->setHeight(STARTING_HEIGHT);
+    newNode->setInfo(info);
+
+    this->root = insertAVLTree(newNode, this->root);
+}
+
+template<class Key, class Info>
+void AVLTree<Key, Info>::insert(shared_ptr<TNode<Key, Info>> newNode) {
+    newNode->setHeight(STARTING_HEIGHT);
+
+    this->root = insertAVLTree(newNode, this->root);
+}
+
+template<class Key, class Info>
+void AVLTree<Key, Info>::remove(const Key &key) {
+    this->root = removeFromAVLTree(this->root, key);
+}
+
+template<class Key, class Info>
+void AVLTree<Key, Info>::printTreeInfoInOrder(std::ostream &os) const {
+    os << "Tree Info summary:" << std::endl;
+
+    printTreeInfoInOrder(os, this->root);
+}
+
+template<class Key, class Info>
+void AVLTree<Key, Info>::treeClear() {
+    this->root = nullptr;
+}
+
+template<class Key, class Info>
+shared_ptr<TNode<Key, Info>> AVLTree<Key, Info>::getNextNodeInOrder(shared_ptr<TNode<Key, Info>> node) const {
+    if (node == nullptr) {
+        return nullptr;
+    } else if (node->getRight() != nullptr) {
+        return getNextNodeInOrderLeft(node->getRight());
+    }
+
+    return getNextNodeInOrderFromRoot(this->root, node, nullptr);
+}
+
+template<class Key, class Info>
+shared_ptr<TNode<Key, Info>>
+AVLTree<Key, Info>::searchBinaryTree(const Key &key, shared_ptr<TNode<Key, Info>> treeNode) const {
+    if (treeNode == nullptr) {
+        return nullptr;
+    }
+
+    if (key == treeNode->getKey()) {
+        return treeNode;
+    } else if (key < treeNode->getKey()) {
+        return searchBinaryTree(key, treeNode->getLeft());
+    } else if (key > treeNode->getKey()) {
+        return searchBinaryTree(key, treeNode->getRight());
+    }
+
+    // Code execution should never get here
+    assert(false);
+
+    return nullptr;
+}
+
+template<class Key, class Info>
+shared_ptr<TNode<Key, Info>> AVLTree<Key, Info>::insertAVLTree(shared_ptr<TNode<Key, Info>> newNode,
+                                                               shared_ptr<TNode<Key, Info>> treeNode) {
+    if (treeNode == nullptr) {
+        return newNode;
+    }
+
+    if (treeNode->getKey() == newNode->getKey()) {
+        throw KeyAlreadyExistsException();
+    } else if (treeNode->getKey() < newNode->getKey()) {
+        treeNode->setRight(insertAVLTree(newNode, treeNode->getRight()));
+    } else if (treeNode->getKey() > newNode->getKey()) {
+        treeNode->setLeft(insertAVLTree(newNode, treeNode->getLeft()));
+    }
+
+    updateHeight(treeNode);
+
+    return fixTreeBalance(treeNode);
```

Comment:
would've added a comment saying "it will be called from leaf to root" or something like that

Reply:
fixxxxxxxxxed

Reply:
where?